File: games/BastionTD/src_python/wave_manager.py
"""
wave_manager.py - Wave definitions, spawn queue, timing, phase control.
"""
from __future__ import annotations

from typing import TYPE_CHECKING
from settings import *
from enemy import Enemy

if TYPE_CHECKING:
    pass  # grid.Grid type hint if needed


class WaveManager:
    """Manages wave progression, enemy spawning, and build/wave phase transitions."""

    def __init__(self) -> None:
        self.waves: list = generate_waves()
        self.current_wave: int = 0          # 0-indexed
        self.phase: str = "build"           # "build" or "wave"
        self.spawn_queue: list[tuple[str, float]] = []  # [(enemy_type, delay)]
        self.active_enemies: list[Enemy] = []
        self.spawn_timer: float = 0.0

    # ------------------------------------------------------------------
    # Start wave
    # ------------------------------------------------------------------
    def start_wave(self) -> None:
        if self.current_wave >= len(self.waves):
            return
        self.phase = "wave"
        wave_def = self.waves[self.current_wave]
        self.spawn_queue = []

        # Build spawn queue: groups spawn sequentially.
        # Each group is (enemy_type, count, spawn_delay).
        # Titan always last (already guaranteed by generate_waves ordering).
        for enemy_type, count, delay in wave_def:
            for i in range(count):
                # First enemy in a group spawns immediately (0 delay), rest use spawn_delay
                entry_delay = 0.0 if i == 0 else delay
                self.spawn_queue.append((enemy_type, entry_delay))

        self.spawn_timer = 0.0

    # ------------------------------------------------------------------
    # Update -- returns list of event dicts
    # ------------------------------------------------------------------
    def update(self, dt: float, grid) -> list[dict]:
        events: list[dict] = []

        if self.phase != "wave":
            return events

        # 1. Spawn enemies from queue -----------------------------------
        if self.spawn_queue:
            self.spawn_timer -= dt
            if self.spawn_timer <= 0:
                enemy_type, delay = self.spawn_queue.pop(0)

                # Pick a path from the grid -- use first available spawn
                path = None
                if grid.paths:
                    # Use the first spawn path (or round-robin for multi-spawn)
                    spawn_keys = list(grid.paths.keys())
                    spawn_key = spawn_keys[len(self.active_enemies) % len(spawn_keys)]
                    path = grid.paths[spawn_key]

                if path:
                    enemy = Enemy(enemy_type, path, spawn_index=0)
                    self.active_enemies.append(enemy)
                    # Boss spawn event
                    if enemy_type == "titan":
                        events.append({
                            "type": "boss_spawn",
                            "enemy": enemy,
                        })

                # Set timer for next spawn
                if self.spawn_queue:
                    _, next_delay = self.spawn_queue[0]
                    self.spawn_timer = next_delay
                else:
                    self.spawn_timer = 0.0

        # 2. Update all active enemies -----------------------------------
        for enemy in self.active_enemies:
            enemy.update(dt, self.active_enemies)

        # 3. Process dead and base-reaching enemies ----------------------
        still_active: list[Enemy] = []
        for enemy in self.active_enemies:
            if enemy.reached_base:
                events.append({
                    "type": "lives_lost",
                    "amount": enemy.lives_cost,
                    "enemy_type": enemy.type,
                })
            elif not enemy.alive:
                # Enemy died (not from reaching base) -- award gold
                if enemy.death_timer <= 0:
                    events.append({
                        "type": "gold_earned",
                        "amount": enemy.gold,
                        "enemy_type": enemy.type,
                        "x": enemy.x,
                        "y": enemy.y,
                    })
                else:
                    # Still in death animation
                    still_active.append(enemy)
            else:
                still_active.append(enemy)

        self.active_enemies = still_active

        # 4. Check wave complete -----------------------------------------
        if not self.spawn_queue and not self.active_enemies:
            events.append({"type": "wave_complete", "wave": self.current_wave})
            self.current_wave += 1
            self.phase = "build"

        return events

    # ------------------------------------------------------------------
    # Queries
    # ------------------------------------------------------------------
    def is_wave_active(self) -> bool:
        return self.phase == "wave"

    def enemies_remaining(self) -> int:
        return len(self.spawn_queue) + len(self.active_enemies)

File: games/BastionTD/src_cpp/core/wave_manager.cpp
#include "core/wave_manager.h"

void WaveManager::init() {
    cfg::generate_waves(wave_defs);
    current_wave = 0;
    phase = GamePhase::Build;
    queue_size = 0;
    queue_head = 0;
    spawn_timer = 0.0f;
}

bool WaveManager::start_wave(int spawn_count) {
    if (phase != GamePhase::Build || current_wave >= cfg::TOTAL_WAVES) {
        return false;
    }

    phase = GamePhase::Wave;
    queue_size = 0;
    queue_head = 0;
    spawn_timer = 0.0f;

    const auto& wave = wave_defs[current_wave];
    for (int entry_idx = 0; entry_idx < wave.entry_count; ++entry_idx) {
        const auto& entry = wave.entries[entry_idx];
        for (int i = 0; i < entry.count && queue_size < 256; ++i) {
            const float delay = (i == 0) ? 0.0f : entry.spawn_delay;
            const int path_choice = spawn_count > 1 ? (queue_size % spawn_count) : 0;
            spawn_queue[queue_size++] = {entry.type, path_choice, delay};
        }
    }

    return true;
}

void WaveManager::update(float dt, EnemyPool& enemies, const Path paths[], int path_count) {
    if (phase != GamePhase::Wave) {
        return;
    }

    spawn_timer -= dt;
    while (queue_head < queue_size && spawn_timer <= 0.0f) {
        const SpawnEntry entry = spawn_queue[queue_head];
        const int safe_path_idx = (path_count > 0) ? (entry.path_idx % path_count) : 0;
        Enemy* spawned = enemies.spawn(entry.type, &paths[safe_path_idx]);
        if (spawned == nullptr) {
            spawn_timer = 0.0f;
            break;
        }
        ++queue_head;
        if (queue_head < queue_size) {
            spawn_timer += spawn_queue[queue_head].delay;
        }
    }
}

bool WaveManager::is_wave_complete(const EnemyPool& enemies) const {
    return phase == GamePhase::Wave && queue_head >= queue_size && enemies.active_count() == 0;
}

int WaveManager::enemies_remaining(const EnemyPool& enemies) const {
    return (queue_size - queue_head) + enemies.active_count();
}

bool WaveManager::all_waves_done() const {
    return current_wave >= cfg::TOTAL_WAVES;
}
